File: seminars/2026-04-05-debugging/main.py
# example.py

def compute_average(data):
    return sum(data) / len(data)        # potential division by zero


def find_name_by_id(idx):
    names = ["Alice", "Bob", "Charlie"]
    return names[idx]               # BUG: no bounds check


def main():
    numbers = [10, 20, 30, 40]

    print("Average:", compute_average(numbers))

    idx = int(input("Enter ID (0-2): "))
    name = find_name_by_id(idx)
    print("Name:", name)

    value = 42
    ref = value
    del value
    print("Value after delete:", ref)  # subtle: still works (good discussion point)


if __name__ == "__main__":
    main()

File: seminars/2026-04-05-debugging/main.cc
#include <iostream>
#include <vector>
#include <string>

int compute_average(const std::vector<int>& data) {
    int sum = 0;
    for (size_t i = 0; i <= data.size(); ++i) {  // BUG: off-by-one (<=)
        sum += data[i];                          // potential out-of-bounds
    }
    return sum / data.size();                    // potential divide by zero
}

std::string find_name_by_id(int id) {
    std::vector<std::string> names = {"Alice", "Bob", "Charlie"};
    return names[id];  // BUG: no bounds checking
}

int main() {
    std::vector<int> numbers = {10, 20, 30, 40};

    std::cout << "Average: " << compute_average(numbers) << std::endl;

    int id;
    std::cout << "Enter ID (0-2): ";
    std::cin >> id;

    std::string name = find_name_by_id(id);
    std::cout << "Name: " << name << std::endl;

    int* ptr = new int(42);
    delete ptr;
    std::cout << "Value after delete: " << *ptr << std::endl; // BUG: use-after-free

    return 0;
}
